providers/implementations/signature/rsa.c

```diff
@@ -0,0 +1,1006 @@
+
+    /* Temp buffer */
+    unsigned char *tbuf;
+
+} PROV_RSA_CTX;
+
+static void *rsa_newctx(void *provctx)
+{
+    PROV_RSA_CTX *prsactx = OPENSSL_zalloc(sizeof(PROV_RSA_CTX));
+
+    if (prsactx == NULL)
+        return NULL;
+
+    prsactx->libctx = PROV_LIBRARY_CONTEXT_OF(provctx);
+    return prsactx;
+}
+
+/* True if PSS parameters are restricted */
+#define rsa_pss_restricted(prsactx) (prsactx->min_saltlen != -1)
+
+static int rsa_signature_init(void *vprsactx, void *vrsa)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+
+    if (prsactx == NULL || vrsa == NULL || !RSA_up_ref(vrsa))
+        return 0;
+
+    RSA_free(prsactx->rsa);
+    prsactx->rsa = vrsa;
+    if (RSA_get0_pss_params(prsactx->rsa) != NULL)
+        prsactx->pad_mode = RSA_PKCS1_PSS_PADDING;
+    else
+        prsactx->pad_mode = RSA_PKCS1_PADDING;
+    /* Maximum for sign, auto for verify */
+    prsactx->saltlen = RSA_PSS_SALTLEN_AUTO;
+    prsactx->min_saltlen = -1;
+
+    return 1;
+}
+
+static int setup_md(PROV_RSA_CTX *ctx)
+{
+    if (ctx->mdname[0] != '\0') {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, ctx->mdname, ctx->mdprops);
+
+        if (md == NULL)
+            return 0;
+        EVP_MD_CTX_free(ctx->mdctx);
+        EVP_MD_free(ctx->md);
+        ctx->md = md;
+        ctx->mdsize = EVP_MD_size(md);
+        ctx->mdctx = EVP_MD_CTX_new();
+        if (ctx->mdctx == NULL)
+            return 0;
+
+        if (!EVP_DigestInit_ex(ctx->mdctx, md, NULL))
+            return 0;
+    }
+
+    if (ctx->mgf1_mdname[0] != '\0') {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, ctx->mgf1_mdname,
+                                  ctx->mgf1_mdprops);
+
+        if (md == NULL)
+            return 0;
+        EVP_MD_free(ctx->mgf1_md);
+        ctx->mgf1_md = md;
+    }
+
+    return 1;
+}
+
+static int setup_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        return 1;
+    if ((ctx->tbuf = OPENSSL_malloc(RSA_size(ctx->rsa))) == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    return 1;
+}
+
+static void clean_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_cleanse(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static void free_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_clear_free(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static int rsa_sign(void *vprsactx, unsigned char *sig, size_t *siglen,
+                    size_t sigsize, const unsigned char *tbs, size_t tbslen)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+    int ret;
+
+    if (sig == NULL) {
+        *siglen = RSA_size(prsactx->rsa);
+        return 1;
+    }
+
+    if (sigsize < (size_t)RSA_size(prsactx->rsa))
+        return 0;
+
+    if (!setup_md(prsactx))
+        return 0;
+
+    if (prsactx->md != NULL) {
+        if (tbslen != prsactx->mdsize) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST_LENGTH);
+            return 0;
+        }
+
+#ifdef LEGACY_MODE
```

Comment:
Huh? Why would we be compiling this code for the legacy provider?? Surely it only ever sits in the FIPS or default providers...even if in the default provider it happens to be using a digest implemented in the legacy provider?

I would have expected this condition to be:

`#if !defined(FIPS_MODE)`

Reply:
Well...  no.  MDC2 is a legacy algorithm, found in the legacy provider, and we *do* support the `no-legacy` configuration.

However, the code in here is implementation agnostic, it only depends on the presence of the macro `OSSL_DIGEST_NAME_MDC2`, so that guard should be completely unnecessary.  It may be that I had a stricter dependence on the MDC2 code at some point, and forgot to remove the guard when it became more agnostic.

Removing the guard now.

Reply:
The reason for the guard was https://github.com/openssl/openssl/pull/10557#pullrequestreview-326547695

This piece of code should not appear in the FIPS provider. It has nothing to do with the legacy provider. The fact that the MDC2 digest is found in the legacy provider is irrelevant (as you say the code here is implementation agnostic). It still seems to me that the correct guard is as I showed above (which I note that @slontis also suggests in the comment above!)

Mind you  - you can argue that we can defer this conversation until we actually pull this code into the FIPS provider.

Reply:
No, I'll add the FIPS_MODE guard.  Thanks for the reminder